[Louis Lie] iP

louisdavinlie opened the pull request:
## Duke

> "If you don't have an assistant, you are one!" - Cameron Herold

Duke is the only personal assistant you ever need. It's,

- Interactive
- Easy to learn
- Text-based
- ~~Fast and easy~~ **_SUPER FAST AND EASY_** to use

To begin using this amazing assistant, you need to:

1. Download the code base from [here](https://github.com/louisdavinlie/ip) ⬇️ 
2. Run the Launcher class 🚀 
3. Add your tasks 📃 
4. Let it make your life simpler! 🤑 

Features:

- [x] Add any task you have.
- [x] Manage your tasks, deadlines and events.
- [x] Find any task you have.
- [ ] Reminders! (Coming soon)
----------------
Below is the `main` method in the Launcher class you need to run,

```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

louisdavinlie committed:
Implement greet, echo, exit

louisdavinlie committed:
Implement add, list

Level-2

louisdavinlie committed:
Add Task.java

louisdavinlie committed:
Implement mark as done

Level-3

louisdavinlie committed:
Add Deadline.java, Event.java, Todo.java

Level-4

louisdavinlie committed:
Implement ToDo, event, deadline

Level-4

louisdavinlie committed:
Add Automated Text UI Testing

louisdavinlie committed:
Add Automated Text UI Testing

louisdavinlie committed:
Merge branch 'master' of https://github.com/louisdavinlie/ip

louisdavinlie committed:
Implement exceptions

Level-5

louisdavinlie committed:
Implement delete

Level-6

louisdavinlie committed:
Implement save

lkldev commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + this.getStatusIcon() + " " + super.description + "(by:" + this.dateAndTime + ")";
```
Following the DRY Principle, instead of using `this.getStatusIcon() + " " + super.description`, I feel that it will be better to use the toString method in the parent class `Task`, since the code is repeated for all 3 types of tasks.
```suggestion
        return "[D]" + super.toString() + "(by:" + this.dateAndTime + ")";
```

lkldev commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,153 @@
+                }
+            } else {
+                throw new DukeException("I don't understand your query.");
+            }
```
I feel that using a switch case here would be neater and easier to read. You can use the "fall through" method to process the same type of command, e.g.
```java
switch (command) {
...
case "mark":
//Fall through
case "unmark":
//Fall through
case "delete":
    if (instructions.length == 1) {
            throw new DukeException("Please specify the task number to " + command + ".");
    }
    int taskNumber = Integer.parseInt(instructions[1]);
    if (taskNumber > taskList.size() || taskNumber <= 0) {
        throw new DukeException("Invalid task number! You have " + taskList.size() + " task(s).");
    }
    break;
...
}
```

lkldev commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,73 @@
+        } catch (IOException e) {
+            System.out.println(e.getMessage());
+        }
+    }
```
Great method to abstract out the file checking process! Makes the code more readable 👍

lkldev commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,153 @@
+        Task task = null;
+        if (instructions[0].equals("todo")) {
+            task = new Todo(instructions[1]);
+            System.out.println("Added a to do task.");
```
I feel that all the printed messages can be replaced with a named constant so that it will be easier to read the code since it is not important as to what the bot would print. An example would be:
```java
private static final String ADD_TODO_MESSAGE = "Added a to do task.";
...
System.out.println(ADD_TODO_MESSAGE);
```
This is easily understood that the bot would print a message. Also, having all the messages consolidated allow for easy changing of the message sent if an update is required.

lkldev commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,153 @@
+        greetings();
+        chat();
+        bye();
     }
```
Great level of abstraction here! it follows the Single Level of Abstraction Principle, making it easy to understand what is the main logic here.

lkldev commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+
+    public void unmark() {
+        this.isDone = false;
+    }
```
Since there is a `markAsDone` method, I feel that this method should be named as `unmarkAsDone` or something similar so that it is easy to understand that these two methods are opposites of each other.

lkldev commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,153 @@
+            String[] instructions = userInput.split(" ", 2);
+            try {
+                checkCommand(instructions);
+                if (instructions[0].equals("bye")) {
+                    storage.writeToFile(taskList);
+                    break;
+                } else if (instructions[0].equals("list")) {
+                    list();
+                } else if (instructions[0].equals("mark") || instructions[0].equals("unmark") || instructions[0].equals("delete")) {
+                    modify(instructions);
+                } else {
+                    add(instructions);
+                }
```
Following the Single Level of Abstraction Principle, since there is a `checkCommand` method, maybe it will be good to have a `processCommand` method as well?

lkldev reviewed (commented):
Overall, good code! There are some nits that I have picked out based on the [code quality guidelines](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html). Generally, great naming of variables and methods, and the code is easy to follow.

louisdavinlie committed:
Add date and time support

branch-Level-8

louisdavinlie committed:
add and modify files to implement more OOP

louisdavinlie committed:
divide classes into packages

louisdavinlie committed:
make directory if not exist and minor change

louisdavinlie committed:
add files for JUnit support

louisdavinlie committed:
add FindCommand.java, implement find function

louisdavinlie committed:
remove extra line when calling invalid command

louisdavinlie committed:
add JavaDoc comments

louisdavinlie committed:
change method names

louisdavinlie committed:
Merge branch 'branch-A-JavaDoc'

louisdavinlie committed:
Merge branch 'branch-A-CodingStandard'

louisdavinlie committed:
Merge branch 'branch-Level-9'

louisdavinlie committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

louisdavinlie committed:
Merge branch 'branch-A-Gradle'

louisdavinlie committed:
add config files

louisdavinlie committed:
change to follow checkstyle

louisdavinlie committed:
Merge branch 'branch-A-CheckStyle'

louisdavinlie committed:
add gui and change methods to fit gui

louisdavinlie committed:
Merge branch 'branch-Level-10'

louisdavinlie committed:
Delete dead codes.

There are several dead codes in the code base since a GUI had
been implemented. Most of them are located in Duke.java and Ui.java.

Since a GUI had been implemented, the way to display Duke's responses
has changed. Instead of using System.out.println(), the string of
response needs to be returned to be displayed on the GUI.

louisdavinlie committed:
Use assertions

Added assertions to some logic so that the runtime can verify
them.

louisdavinlie committed:
Merge pull request #2 from louisdavinlie/branch-A-CodeQuality

Delete dead codes

louisdavinlie committed:
Merge branch 'master' into branch-A-Assertions

louisdavinlie committed:
Merge pull request #3 from louisdavinlie/branch-A-Assertions

Use assertions

louisdavinlie committed:
Add help command

louisdavinlie committed:
Modify GUI to have a better look

Increased the width and height of Duke window. Added Duke title.
Exit application when user inputs "bye". Resized display picture.
Changed font text.

louisdavinlie committed:
Change FXML styling

louisdavinlie committed:
Merge branch 'branch-A-BetterGui'

louisdavinlie committed:
Add product screenshot (Ui.png)

louisdavinlie committed:
Update README.md

louisdavinlie committed:
Set theme jekyll-theme-slate

louisdavinlie committed:
Modify gradle mainClassName

louisdavinlie committed:
Merge branch 'master' of https://github.com/louisdavinlie/ip

louisdavinlie committed:
Add JavaDoc comment for Duke getResponse() method
